### DISCUSSION TITLE
Signal old container to do cleanup?

### DISCUSSION THREAD

**Discussion by GormanFletcher** [dokku/dokku#4708] [Q&A]
When my app gets deployed, is there a way to signal the old container to do internal cleanup during the `DOKKU_WAIT_TO_RETIRE` interval? I'd like to e.g., stop taking tasks from the queue once the container knows termination is imminent.

**Comment by josegonzalez** [comment]
We run `docker stop` prior to stopping the container (this is currently controlled by `DOKKU_DOCKER_STOP_TIMEOUT`). This sends a SIGTERM and then a SIGKILL. Your process should respect SIGTERM and cleanup as appropriate.

**Reply by GormanFletcher** [reply]
Can my container detect when it's in the `DOKKU_WAIT_TO_RETIRE` period? I'd like my container to have a chance to complete in-progress work during `DOKKU_WAIT_TO_RETIRE`, but the container needs to know not to take new work during that period. If the container doesn't know it's being cycled out until the `SIGTERM` arrives, I have to ignore `DOKKU_WAIT_TO_RETIRE` and use `DOKKU_DOCKER_STOP_TIMEOUT` for the purpose the docs give for `DOKKU_WAIT_TO_RETIRE`.

**Reply by josegonzalez** [reply]
At the moment, there isn't a great way. We could send a different (what would that be? `SIGQUIT` seems closest) signal, though afaik Heroku doesn't (though they do have a 30 second timeout instead of our default 10), and I'd like to not deviate from this if at all possible.

It might be nice to have custom `DOKKU_WAIT_TO_RETIRE` values on a per-process basis, so you could theoretically set that to `0` and then have the `DOKKU_DOCKER_STOP_TIMEOUT` be 30 (or whatever) for a process. That way, we'd immediately attempt to retire the process, and the `docker container stop` call would give it enough time to do whatever it needs finish processing.

**Comment by sixers** [comment]
I'm also interested in this feature. Is it possible to implement this via plugin?

**Reply by josegonzalez** [reply]
The SIGTERM signal should be caught by your app and used to do cleanup before exiting cleanly. Is there a reason this doesn't suffice?
